lib/livebook/runtime/node_pool.ex

```diff
@@ -0,0 +1,102 @@
+defmodule Livebook.Runtime.NodePool do
+  use GenServer
+
+  @moduledoc false
+
+  # A pool for reusing child node names.
+  #
+  # `own_pool` refers to the list of unused names.
+  # `given_pool` refers to the list of names ever generated.
+  #
+  # `buffer_time` refers to time the pool waits before
+  # adding a name to `pool`,  which by default is 1 minute.
+
+  @default_time 60_000
+
+  # Client interface
+
+  @doc """
+  Starts the GenServer from a Supervision tree
+  """
+  def start_link(opts) do
+    # Note that the following options are for testing only.
+    name = opts[:name] || __MODULE__
+    buffer_time = opts[:buffer_time] || @default_time
+    given_pool = opts[:given_pool] || []
+    own_pool = opts[:own_pool] || []
+
+    GenServer.start_link(
+      __MODULE__,
+      %{buffer_time: buffer_time, given_pool: given_pool, own_pool: own_pool},
+      name: name
+    )
+  end
+
+  @doc """
+  Retuns a node name.
+
+  Generates a new name if pool is empty, or takes one from pool.
+  """
+  def get_name(basename, server \\ __MODULE__) do
```

Comment:
Nit: I think it's a convention to always pass the server name/pid as the first argument (and you can still use a default value :>)